File: util/java/src/main/java/com/dmi/util/cache/ReuseCache.java
package com.dmi.util.cache;

import com.carrotsearch.hppc.ByteArrayList;
import com.carrotsearch.hppc.CharArrayList;
import com.carrotsearch.hppc.FloatArrayList;
import com.carrotsearch.hppc.IntArrayList;
import com.dmi.util.annotation.Reusable;

import java.lang.ref.WeakReference;
import java.util.Collection;
import java.util.concurrent.atomic.AtomicReference;

import javax.annotation.concurrent.NotThreadSafe;

import java8.util.function.Supplier;

@NotThreadSafe
public abstract class ReuseCache {
    public static <T> Reuser<T> reuser(Supplier<T> creator) {
        AtomicReference<WeakReference<T>> holder = new AtomicReference<>(new WeakReference<>(null));
        return () -> {
            WeakReference<T> weakRef = holder.get();
            T value = weakRef.get();
            if (value == null) {
                value = creator.get();
                holder.set(new WeakReference<>(value));
            }
            return value;
        };
    }

    public static StringBuilder reuseStringBuilder(Reuser<StringBuilder> reuser) {
        StringBuilder value = reuser.reuse();
        value.setLength(0);
        return value;
    }

    public static <T extends Collection<?>> T reuseCollection(Reuser<T> reuser) {
        T value = reuser.reuse();
        value.clear();
        return value;
    }

    public static IntArrayList reuseIntArrayList(Reuser<IntArrayList> reuser) {
        IntArrayList value = reuser.reuse();
        value.elementsCount = 0;
        return value;
    }

    public static ByteArrayList reuseByteArrayList(Reuser<ByteArrayList> reuser) {
        ByteArrayList value = reuser.reuse();
        value.elementsCount = 0;
        return value;
    }

    public static FloatArrayList reuseFloatArrayList(Reuser<FloatArrayList> reuser) {
        FloatArrayList value = reuser.reuse();
        value.elementsCount = 0;
        return value;
    }

    public static CharArrayList reuseCharArrayList(Reuser<CharArrayList> reuser) {
        CharArrayList value = reuser.reuse();
        value.elementsCount = 0;
        return value;
    }

    public static float[] reuseFloatArray(Reuser<AtomicReference<float[]>> reuser, int capacity) {
        AtomicReference<float[]> ref = reuser.reuse();
        float[] value = ref.get();
        if (value.length < capacity) {
            value = new float[capacity + capacity >> 1];
            ref.set(value);
        }
        return value;
    }

    public static boolean[] reuseBooleanArray(Reuser<AtomicReference<boolean[]>> reuser, int capacity) {
        AtomicReference<boolean[]> ref = reuser.reuse();
        boolean[] value = ref.get();
        if (value.length < capacity) {
            value = new boolean[capacity + capacity >> 1];
            ref.set(value);
        }
        return value;
    }

    public interface Reuser<T> {
        @Reusable
        T reuse();
    }
}

File: util/util-common/src/main/kotlin/com/dmi/util/cache/ReuseCache.kt
package com.dmi.util.cache

import com.carrotsearch.hppc.ByteArrayList
import com.carrotsearch.hppc.CharArrayList
import com.carrotsearch.hppc.FloatArrayList
import com.carrotsearch.hppc.IntArrayList
import com.dmi.util.annotation.Reusable
import java.lang.ref.WeakReference
import java.util.concurrent.atomic.AtomicReference

object ReuseCache {
    fun <T> reuser(create: () -> T): Reuser<T> {
        val holder = AtomicReference(WeakReference<T>(null))
        return object: Reuser<T> {
            override fun reuse(): T {
                val weakRef = holder.get()
                var value = weakRef.get()
                if (value == null) {
                    value = create()
                    holder.set(WeakReference(value))
                }
                return value
            }
        }
    }

    fun reuseStringBuilder(reuser: Reuser<StringBuilder>): StringBuilder {
        val value = reuser.reuse()
        value.setLength(0)
        return value
    }

    fun <T : MutableCollection<*>> reuseCollection(reuser: Reuser<T>): T {
        val value = reuser.reuse()
        value.clear()
        return value
    }

    fun reuseIntArrayList(reuser: Reuser<IntArrayList>): IntArrayList {
        val value = reuser.reuse()
        value.elementsCount = 0
        return value
    }

    fun reuseByteArrayList(reuser: Reuser<ByteArrayList>): ByteArrayList {
        val value = reuser.reuse()
        value.elementsCount = 0
        return value
    }

    fun reuseFloatArrayList(reuser: Reuser<FloatArrayList>): FloatArrayList {
        val value = reuser.reuse()
        value.elementsCount = 0
        return value
    }

    fun reuseCharArrayList(reuser: Reuser<CharArrayList>): CharArrayList {
        val value = reuser.reuse()
        value.elementsCount = 0
        return value
    }

    fun reuseFloatArray(reuser: Reuser<AtomicReference<FloatArray>>, capacity: Int): FloatArray {
        val ref = reuser.reuse()
        var value = ref.get()
        if (value.size < capacity) {
            value = FloatArray(capacity + capacity shr 1)
            ref.set(value)
        }
        return value
    }

    fun reuseBooleanArray(reuser: Reuser<AtomicReference<BooleanArray>>, capacity: Int): BooleanArray {
        val ref = reuser.reuse()
        var value = ref.get()
        if (value.size < capacity) {
            value = BooleanArray(capacity + capacity shr 1)
            ref.set(value)
        }
        return value
    }

    interface Reuser<T> {
        @Reusable
        fun reuse(): T
    }
}
